src/overlays/actors/ovl_Bg_Goron_Oyu/z_bg_goron_oyu.c

```diff
@@ -28,31 +34,163 @@ const ActorInit Bg_Goron_Oyu_InitVars = {
+
+    if (this->unk_164 <= 0.0f) {
+        ActorCutscene_Stop(this->initialActorCutscene);
+        this->unk_164 = 0.0f;
+        func_80B40080(this);
+    }
+
+    func_8019F1C0(&D_80B40780, NA_SE_EV_WATER_LEVEL_DOWN - SFX_FLAG);
+}
+
+void func_80B401F8(BgGoronOyu* this, GlobalContext* globalCtx) {
+    Player* player;
+    Vec3f dist;
+
+    if (Actor_HasParent(&this->dyna.actor, globalCtx)) {
+        this->dyna.actor.parent = NULL;
+        return;
+    }
+
+    player = GET_PLAYER(globalCtx);
+    Math_Vec3f_DistXYZAndStoreDiff(&this->waterBoxPos, &player->actor.world.pos, &dist);
+
+    if (dist.x >= 0.0f && dist.x <= this->waterBoxXLength && dist.z >= 0.0f && dist.z <= this->waterBoxZLength &&
+        fabsf(dist.y) < 100.0f && player->actor.depthInWater > 12.0f) {
+        func_800B8A1C(&this->dyna.actor, globalCtx, 0xBA, this->dyna.actor.xzDistToPlayer,
+                      fabsf(this->dyna.actor.playerHeightRel));
+    }
+}
+
+void BgGoronOyu_UpdateWaterBoxInfo(BgGoronOyu* this, GlobalContext* globalCtx) {
+    WaterBox* waterBox;
+    f32 ySurface;
+
+    if (WaterBox_GetSurface1(globalCtx, &globalCtx->colCtx, this->dyna.actor.world.pos.x, this->dyna.actor.world.pos.z,
+                             &ySurface, &waterBox)) {
+        Math_Vec3s_ToVec3f(&this->waterBoxPos, &waterBox->minPos);
+        this->waterBoxXLength = waterBox->xLength;
+        this->waterBoxZLength = waterBox->zLength;
+    }
+}
+
+void BgGoronOyu_SpawnParticles(BgGoronOyu* this, GlobalContext* globalCtx) {
+    s16 scale;
+    Vec3f pos1;
+    Vec3f pos2;
+    CollisionPoly* poly;
+    s32 pad;
+
+    if ((globalCtx->state.frames % 4) == 0) {
+        Vec3f vel1;
+        Vec3f vel2;
+
+        pos1.x = this->dyna.actor.world.pos.x;
+        pos1.y = this->dyna.actor.world.pos.y + 100.0f;
+        pos1.z = this->dyna.actor.world.pos.z;
+        vel1.x = 0.0f;
+        vel1.y = 2.5f;
+        vel1.z = 0.0f;
+        scale = -200 - (s32)(Rand_ZeroOne() * 50.0f);
+
+        if (BgCheck_EntityRaycastFloor2(globalCtx, &globalCtx->colCtx, &poly, &pos1) < this->waterBoxPos.y) {
+            pos1.y = this->waterBoxPos.y + 10.0f;
+            EffectSsIceSmoke_Spawn(globalCtx, &pos1, &vel1, &D_801D15B0, scale);
+        }
+        pos2.x = (Rand_ZeroOne() * this->waterBoxXLength) + this->waterBoxPos.x;
+        pos2.y = this->waterBoxPos.y + 100.0f;
+        pos2.z = (Rand_ZeroOne() * this->waterBoxZLength) + this->waterBoxPos.z;
+        vel2.x = 0.0f;
+        vel2.y = 0.5f;
+        vel2.z = 0.0f;
+        scale = -200 - (s32)(Rand_ZeroOne() * 50.0f);
+
+        if (BgCheck_EntityRaycastFloor2(globalCtx, &globalCtx->colCtx, &poly, &pos2) < this->waterBoxPos.y) {
+            pos2.y = this->waterBoxPos.y + 10.0f;
+            EffectSsIceSmoke_Spawn(globalCtx, &pos2, &vel2, &D_801D15B0, scale);
+        }
+    }
+}
+
+void BgGoronOyu_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgGoronOyu* this = THIS;
+    s32 pad;
```

Comment:
```suggestion
    BgGoronOyu* this = THIS;
    s32 pad;
```
If this is a GlobalContext upcast pad, it tends to go above the Actor recast.